Add wrappers for `SO_PEERCRED` and `SCM_CREDENTIALS`

ids1024 opened the pull request:
I still need to test this to make  sure it works properly. Particularly the `SCM_CREDENTIALS` part.

Apparently [FreeBSD](https://man.freebsd.org/cgi/man.cgi?query=unix&sektion=4), macOS, and DragonFlyBSD have `LOCAL_PEERCRED` which similar to `SO_PEERCRED` but with an array of groups. And likewise FreeBSD lists `SCM_CREDS` with a similar difference... There's also `getpeereid` on all BSDs but not Linux that gets uid and gid but not pid.

sunfishcode commented:
On the merge conflict here, #855 moved all the sockopt functions out of syscalls.rs and into their own sockopt.rs file, so hopefully you can just move the code here over to the new file.

ids1024 commented:
I was thinking this would be awkward to test, but actually it's not to hard to just create a `socketpair`, and there's no need for the socket ends to be in different processes. So this has a test now showing these APIs work and return the correct values.

I also updated `UCred` to use the `Pid`/`Uid`/`Gid` types instead of raw integers, and updated the `cmsg_space!` macro to handle `ScmCredentials`.

sunfishcode reviewed (commented):
Thanks for working on this!

ids1024 commented on src/lib.rs:
```diff
@@ -335,7 +335,8 @@ mod clockid;
     feature = "thread",
-    all(bsd, feature = "event")
+    all(bsd, feature = "event"),
+    all(target_os = "linux", any(target_os = "android", target_os = "linux"))
```
Ah right, I think I meant to use `feature = "net"` here.

ids1024 commented on src/net/send_recv/msg.rs:
```diff
@@ -316,6 +338,28 @@ impl<'buf> AncillaryDrain<'buf> {
+                            gid: Gid::from_raw(u32::from_ne_bytes(
+                                payload[8..12].try_into().unwrap(),
+                            )),
+                        };
```
Yeah, after updating it to use `Gid`/`Uid`/`Pid` it's even more verbose, and won't just be optimized away. A `read_unaligned` call should be fine.

sunfishcode commented:
Looks good!

sunfishcode commented:
This is now released in rustix 0.38.17.
